### PR TITLE
[SDK] Add onClose callback to details modal

### PR DESCRIPTION
**Pull request by kien-ngo**
CNCT-2444
CNCT-2444

---

## PR-Codex overview
This PR focuses on enhancing the `ConnectButton` component by adding an `onClose` callback for the details modal, improving functionality and user experience. It also includes various updates to the `ChainIcon`, `NetworkSwitcherButton`, and related components.

### Detailed summary
- Added `onClose` callback to `ConnectButtonProps` for modal closure.
- Updated `ConnectButton` documentation to include usage of `onClose`.
- Enhanced `NetworkSwitcherButton` to utilize the new `onClose` callback.
- Improved test coverage for `NetworkSwitcherButton` and related components.
- Refactored `ChainIcon` and `fetchChainIcon` for better icon resolution.
- Added internal documentation for several components.
- Fixed issues related to rendering chain icons and fallback images.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by kien-ngo** [comment]
* **#5605** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/js/5605?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/thirdweb-dev/js/5605?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/thirdweb/src/react/web/ui/ConnectWallet/Details.tsx [hunk 1]
```diff
@@ -990,6 +990,9 @@ function DetailsModal(props: {
             setOpen={(_open) => {
               if (!_open) {
                 closeModal();
+                if (props.detailsModal?.onClose) {
+                  props.detailsModal?.onClose(screen);
+                }
               }
```
[the remark this answers is not in the document]
Comment (kien-ngo) [reply packages/thirdweb/src/react/web/ui/ConnectWallet/Details.tsx:996]: what

**Review by gregfromstl — APPROVED** [review]
LGTM but I'm a little confused how this solves the original issue, also why we need the `screen` param to be passed to `onClose` (feels like something a little too internal to be exposing). I trust though

**Review by joaquim-verges — APPROVED** [review]

**Comment by kien-ngo** [comment]
### Merge activity

* **Dec 6, 9:44 PM EST**: [Graphite](https://app.graphite.dev/github/pr/thirdweb-dev/js/5605) disabled "merge when ready" on this PR due to: a merge conflict with the target branch; resolve the conflict and try again..

File: .changeset/metal-mails-ring.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+---
+"thirdweb": patch
+---
+
+- Add onClose callback to Connect Details modal
+
+```tsx
+<ConnectButton
+  detailsModal={{
+    onClose: (screen: string) => {
```
Comment (gregfromstl) [inline .changeset/metal-mails-ring.md:10]: Rather than burying this in the `detailsModal` prop, could we make it top level?

**Review by gregfromstl — CHANGES_REQUESTED** [review]
Remove screen param at least

File: .changeset/metal-mails-ring.md (hunk 2 above)
Comment (kien-ngo) [reply .changeset/metal-mails-ring.md:10]: This is only applicable for the Modal and not the whole button, so it contextually makes sense to be nested under "detailsModal".

However it took a while for me to raise the coverage to 80% and I'm willing to do what you said to merge this PR lol

Comment (kien-ngo) [reply .changeset/metal-mails-ring.md:10]: Hmm @gregfromstl actually it's cleaner if we keep it inside the `detailsModal` prop. Also by putting onClose at the root level it crashes with the `onClose` of the react native's ConnectButton. (packages/thirdweb/src/react/native/ui/connect/ConnectButton.tsx L114 )

So I'd say we keep things as is. Would you be able to compromise?

**Review by gregfromstl — APPROVED** [review]

**Pull request merged**